compiler/lib/src/main/scala/codegen/XmlWriter/ComputeXmlFiles.scala

```diff
@@ -50,13 +50,13 @@ object ComputeXmlFiles extends AstStateVisitor {
   override def defModuleAnnotatedNode(
     s: State,
     node: Ast.Annotated[AstNode[Ast.DefModule]]
-  ) = {
+  ): Result = {
     val (_, node1, _) = node
     val data = node1.data
     visitList(s, data.members, matchModuleMember)
   }
 
-  override def defPortAnnotatedNode(s: State, node: Ast.Annotated[AstNode[Ast.DefPort]]) = {
+  override def defPortAnnotatedNode(s: State, node: Ast.Annotated[AstNode[Ast.DefPort]]): Either[CodeGenError.DuplicateXmlFile,XmlWriterState] = {
```

Comment:
I'm wondering about this change and others like it. It looks like the migration tool has inferred a specific type and added it as the return type. But this coding style seems like a pain to write by hand. The style I was using before is

* Make a type alias for the more general type, e.g,. type `Result = Either[Error,State]`
* Avoid writing the return type at all where possible
* Otherwise use the type alias

In the new style, it seems to use more specific and more complex instantiations of parameterized types. Is this just an artifact of the migration, or will Scala 3 require this more specific type annotation going forward? And will we have to write these types by hand? Again, this seems like it could be a pain.

Reply:
Completely agree with you... Let me see if I can use the type alias instead.

Reply:
OK, sounds good. If Scala 3 is just being more fussy about writing an explicit return type, maybe we can copy over the return type from the method being overridden? Except in rare cases like the one you identified with the structType, I think this is what is intended and should work.

Reply:
Regarding type aliases, why are there two different aliases for the same type?

```
trait AstStateVisitor extends AstVisitor {

  type State

  type In = State

  type Out = Result.Result[State]

  type Result = Result.Result[State]
```

Reply:
FYI: I think it would be preferable to avoid synonym aliases because it makes the code harder to read.
e.g:

AstVisitor has:

```
def transUnit(in: In, tu: Ast.TransUnit): Out
```

and ComputeXmlFiles, a subtype of AstVisitor has:

```
override def transUnit(s: State, tu: Ast.TransUnit): Result
```

Since `Out` and `Result` are synonyms for the same type, the compiler is OK; however, it creates unnecessary confusion.

Regarding eliminating the return types, unfortunately, the presence of type aliases is potentially confusing.
Is `Either[CodeGenError.DuplicateXmlFile, XmlWriterState]` the appropriate return type or should it be `Out` or `Result`?
All 3 are acceptable to the compiler; for a human, I think the intended return type would help with code legibility, particularly when there is only 1 sensible type alias -- e.g. `Out` or `Result` but not both!

Reply:
FYI:

The IDE find 24 usages of `AstStateVisitor.Out` and 113 usages of `AstStateVisitor.Result`.
I propose refactoring all 24 usages of `AstStateVisitor.Out` as `AstStateVisitor.Result`.

@bocchino Is this OK with you?

Reply:
The rationale for having both is that an AstVisitor is a generic operation that take an In to an Out, whereas an AstStateVisitor manages state. Also, a Result is the result of an operation that can return an error.

So I'd prefer to keep In/Out for AstVisitor and State/Result for AstStateVisitor.

Refactoring uses of AstStateVisitor.Out to AstStateVisitor.Result seems good. We still need to keep the type alias AstStateVisitor.Out to satisfy the inheritance from AstVisitor. Overall I think having synonyms for the same type reflecting its use in different contexts is OK.

Reply:
In English, the logic is something like this: "AstVisitor takes an In to an Out, so to inherit from it you have to define an In type and an out type. AstStateVisitor takes a State to a Result, so it defines those types. It also inherits from AstVisitor, and vis a vis AstVisitor, its State type is the In type of AstVisitor, and its Result type is the Out type of AstVisitor. So we create those type aliases to satisfy the inheritance contract with AstVisitor."